man/download-ensembl-genome.1.ronn
# download-ensembl-genome(1) - download ensembl genome

## SYNOPSIS

`download-ensembl-genome`
`--organism`=<organism>
[`--genome-build`=<genome_build>]
[`--release`=<release>]
[`--output-dir`=<output_dir>]
[`--decompress`]

## DESCRIPTION

Download Ensembl genome and corresponding annotation files.

Creates a new directory in the current working directory by default.

**Important!** When working in R, be sure to match the Ensembl release version
to the latest version supported by AnnotationHub.

The legacy Homo sapiens GRCh37 (hg19) genome build is supported.
If requested, the release version is set internally automatically.

Verify checksums in shell with:

    $ grep "$(sum *.gz)" "CHECKSUMS"

## OPTIONS

  * `--organism`=<organism>:
    Latin name (e.g. `"Homo sapiens"`).

  * `--genome-build`=<genome_build>:
    Genome build (e.g. `"GRCh38"`).
    Defaults to current genome build.

  * `--release`=<release>:
    Ensembl release version (e.g. `100`).
    Defaults to current release.

  * `--output-dir`=<output_dir>:
    Output directory.
    Defaults to working directory.

## SEE ALSO

- https://useast.ensembl.org/
- ftp://ftp.ensembl.org/pub/

Which Homo sapiens genome assembly FASTA to use?

- https://steinbaugh.com/posts/bcbio-homo-sapiens-genome.html
- https://lh3.github.io/2017/11/13/which-human-reference-genome-to-use
- https://bioinformatics.stackexchange.com/questions/540/

Using STAR aligner for RNA-seq?
Here's a relevant section from the manual:
https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf

    ## 2.2.1 Which chromosomes/scaffolds/patches to include?
    ## Examples of acceptable genome sequence files:
    ## - ENSEMBL: files marked with '.dna.primary.assembly'.
    ## - GENCODE: files marked with PRI (primary).

GRCh38 alternatives to Ensembl primary assembly:

- GENCODE assembly.
- ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/
      GCA_000001405.15_GRCh38/
      seqs_for_alignment_pipelines.ucsc_ids/
      GCA_000001405.15_GRCh38_no_alt_analysis_set.fna.gz
- ftp://ftp.ncbi.nlm.nih.gov/genbank/genomes/Eukaryotes/
      vertebrates_mammals/Homo_sapiens/GRCh38/seqs_for_alignment_pipelines/
      GCA_000001405.15_GRCh38_full_plus_hs38d1_analysis_set.fna
- http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/
      GRCh38_reference_genome/GRCh38_full_analysis_set_plus_decoy_hla.fa

GRCh37 alternatives to Ensembl primary assembly:

- ftp://ftp-trace.ncbi.nih.gov/1000genomes/ftp/technical/reference/
      human_g1k_v37.fasta.gz

Note that Ensembl GTF is missing annotations for some transcripts defined
in transcriptome FASTA (human and mouse).
In this case, use '*.chr_patch_hapl_scaff.gtf.gz' GTF file for annotations.

See related issues:

- https://github.com/COMBINE-lab/salmon/issues/122
- https://github.com/COMBINE-lab/salmon/issues/336

## NOTE

R function, defined in r-acidgenomes package.


NEWS
## koopa 0.3.2 (2019-04-18)

- Improved build scripts, and hardened some intended for RedHat Linux.
- Updated genome build information.
- Improved bcbio scripts and custom genome builds.
- New exported scripts: `ip-address.sh`, `vim-sort.sh`.
- Added script to build Python from source for RedHat.



## koopa 0.3.1 (2019-03-19)

- Updated to Ensembl release to 95.
- Updated GENCODE to release human 29 / mouse M20.
- Updated FlyBase to 2019_01.
- `git-pull-all.sh`: Added `git fetch` call.
- Improved build configuration scripts: emacs, git, gsl, openssl, tmux.
- Added bcbio system configuration files (bcbio_system.yaml).
- Improved default run template for bcbio RNA-seq.



## koopa 0.3.0 (2019-02-20)

This release overhauls the previous activation method. Refer to the README
for details on how to reconfigure your shell profile.

- Changed activation method to use `activate.sh` script at top of repo.
- Improved support for ksh and zsh shells. Added Travis CI checks.
- Added native support for zsh pure prompt, including async script. These files
  are saved in `zsh/fpath/`.
- Reorganized bcbio configuration scripts and templates.
- Completely overhauled system activation method. Additional shell
  configuration, including prompt string customization is now disabled by
  default. Refer to the installation instructions for details on how to enable
  this new optional `extra` mode.
- Reorganized installer scripts exported in `bin/` including conda scripts,
  `reset-permissions.sh`, `sudo-yum-update.sh`, etc.



## koopa 0.2.8 (2019-02-12)

- Migrated darwin (macOS)-specific scripts from bash repo to koopa.
- Migrated Harvard O2 and Harvard Odyssey HPC scripts here from bash repo.
- Added `rsync-azure-files.sh` script for working with Azure Shares (Samba).
- Added `autopad-samples.sh` script.
- Added `bash-strict-mode.sh` script.
- Added `md5sum.sh` script.
- Reorganized other scripts accessible via `bin/`.
- Reorganized activation and system scripts, in preparation for 0.3 release.



## koopa 0.2.7 (2019-02-04)

- Improved shellchecks on workflows.
- Split out some activation steps into separate scripts.



## koopa 0.2.6 (2019-01-23)

- Added back ".sh" extension for all exported scripts.
- Updated Travis CI configuration to use shellcheck.
- Reorganized and improved workflow scripts.



## koopa 0.2.5 (2019-01-07)

- Updated installation instructions to recommend cloning to `~/.koopa`.
- Conda now will only activate when `CONDA_DEFAULT_ENV` is set in environment.
  This improves handling in situations for user accounts where we don't want
  to activate conda in a shared environemnt (e.g. bioinfo account on Azure).



## koopa 0.1.1 (2018-09-05)

- Now exporting `KOOPA_VERSION` and `KOOPA_DATE` global variables.
- Updated installation instructions.
- Removed `/setup/` config files.
- Added support for automatic SSH key import, using `$SSH_KEY` variable.
- Removed "g*" aliases for GitHub commands (e.g. ga, gc, gd). Too confusing.
- Simplified the separator bar for koopa message, avoiding unicode.
- Renamed `bcl2fastq_indrop` script to `bcl2fastq_indrops`.
- Added some utility scripts from macOS bash repo:
      - delete_fs_cruft
      - git_pull_all
      - gzip_dir
      - install_tex
      - move_files_up_1_level
      - rename_from_csv
      - reset_permissions



## koopa 0.1.0 (2018-08-24)

- Initial stable release.
- Repo was previously named "seqcloud".


lang/python/include/download-ensembl-genome.py
#!/usr/bin/env python3
"""
Download Ensembl genome.
"""

from argparse import ArgumentParser
from os.path import join, realpath

from koopa.arg import dir_path
from koopa.genome import tx2gene_from_fasta
from koopa.genome.ensembl import (
    download_ensembl_genome,
    download_ensembl_gff,
    download_ensembl_gtf,
    download_ensembl_transcriptome,
    ensembl_version,
)
from koopa.goalie import stop
from koopa.strings import paste_url
from koopa.syntactic import kebab_case
from koopa.system import koopa_help

koopa_help()


parser = ArgumentParser()
parser.add_argument("--organism", required=True, type=str)
parser.add_argument("--build", required=True, type=str)
parser.add_argument("--release", type=str)
parser.add_argument(
    "--type",
    default="all",
    const="all",
    nargs="?",
    choices=["all", "genome", "transcriptome", "none"],
)
parser.add_argument(
    "--annotation",
    default="all",
    const="all",
    nargs="?",
    choices=["all", "gtf", "gff", "none"],
)
parser.add_argument("--output-dir", type=dir_path, default=".")
parser.add_argument("--decompress", action="store_true")
args = parser.parse_args()


def main(
    annotation, build, decompress, genome_type, organism, output_dir, release
):
    """
    Download Ensembl genome.
    """
    if genome_type == "none" and annotation == "none":
        stop("'genome_type' or 'annotation' are required.")
    organism = organism.replace(" ", "_")
    base_url = "ftp://ftp.ensembl.org/pub"
    grch37_base_url = paste_url(base_url, "grch37")
    grch37_release = "87"
    if build == "GRCh37":
        base_url = grch37_base_url
        release = grch37_release
    if release is None:
        release = ensembl_version()
    release_url = paste_url(base_url, "release-" + release)
    output_basename = kebab_case(
        organism + " " + build + " " + "ensembl" + " " + release
    )
    output_dir = join(realpath(output_dir), output_basename)
    if genome_type == "genome":
        download_ensembl_genome(
            organism=organism,
            build=build,
            release_url=release_url,
            output_dir=output_dir,
            decompress=decompress,
        )
    elif genome_type == "transcriptome":
        download_ensembl_transcriptome(
            organism=organism,
            build=build,
            release_url=release_url,
            output_dir=output_dir,
            decompress=decompress,
        )
    elif genome_type == "all":
        download_ensembl_genome(
            organism=organism,
            build=build,
            release_url=release_url,
            output_dir=output_dir,
            decompress=decompress,
        )
        download_ensembl_transcriptome(
            organism=organism,
            build=build,
            release_url=release_url,
            output_dir=output_dir,
            decompress=decompress,
        )
    if annotation == "gtf":
        download_ensembl_gtf(
            organism=organism,
            build=build,
            release=release,
            release_url=release_url,
            output_dir=output_dir,
            decompress=decompress,
        )
    elif annotation == "gff":
        download_ensembl_gff(
            organism=organism,
            build=build,
            release=release,
            release_url=release_url,
            output_dir=output_dir,
            decompress=decompress,
        )
    elif annotation == "all":
        download_ensembl_gtf(
            organism=organism,
            build=build,
            release=release,
            release_url=release_url,
            output_dir=output_dir,
            decompress=decompress,
        )
        download_ensembl_gff(
            organism=organism,
            build=build,
            release=release,
            release_url=release_url,
            output_dir=output_dir,
            decompress=decompress,
        )
    if type != "genome":
        tx2gene_from_fasta(source_name="ensembl", output_dir=output_dir)
    print("Genome downloaded successfully to '" + output_dir + "'.")


main(
    annotation=args.annotation,
    build=args.build,
    decompress=args.decompress,
    genome_type=args.type,
    organism=args.organism,
    output_dir=args.output_dir,
    release=args.release,
)
